src/sync/chase_lev.rs

```diff
@@ -1,602 +0,0 @@
-// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
-// file at the top-level directory of this distribution and at
-// http://rust-lang.org/COPYRIGHT.
-//
-// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
-// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
-// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
-// option. This file may not be copied, modified, or distributed
-// except according to those terms.
-
-//! A lock-free concurrent work-stealing deque
-//!
-//! This module contains a hybrid implementation of the Chase-Lev work stealing deque
-//! described in ["Dynamic Circular Work-Stealing Deque"][chase_lev] and the improved version
-//! described in ["Correct and Efficient Work-Stealing for Weak Memory Models"][weak_chase_lev].
-//! The implementation is heavily based on the pseudocode found in the papers.
-//!
-//! # Example
-//!
-//! ```
-//! use crossbeam::sync::chase_lev;
-//! let (mut worker, stealer) = chase_lev::deque();
-//!
-//! // Only the worker may push/try_pop
-//! worker.push(1);
-//! worker.try_pop();
-//!
-//! // Stealers take data from the other end of the deque
-//! worker.push(1);
-//! stealer.steal();
-//!
-//! // Stealers can be cloned to have many stealers stealing in parallel
-//! worker.push(1);
-//! let stealer2 = stealer.clone();
-//! stealer2.steal();
-//! ```
-//!
-//! [chase_lev]: http://neteril.org/~jeremie/Dynamic_Circular_Work_Queue.pdf
-//! [weak_chase_lev]: http://www.di.ens.fr/~zappa/readings/ppopp13.pdf
-
-use std::cell::UnsafeCell;
-use std::fmt;
-use std::mem;
-use std::ptr;
-use std::sync::atomic::Ordering::{Acquire, Relaxed, Release, SeqCst};
-use std::sync::atomic::{AtomicIsize, fence};
-use std::sync::Arc;
-
-use mem::epoch::{self, Atomic, Shared, Owned};
-
-// Once the queue is less than 1/K full, then it will be downsized. Note that
-// the deque requires that this number be less than 2.
-const K: isize = 4;
-
-// Minimum number of bits that a buffer size should be. No buffer will resize to
-// under this value, and all deques will initially contain a buffer of this
-// size.
-//
-// The size in question is 1 << MIN_BITS
-const MIN_BITS: u32 = 7;
-
-#[derive(Debug)]
-struct Deque<T> {
-    bottom: AtomicIsize,
-    top: AtomicIsize,
-    array: Atomic<Buffer<T>>,
-}
-
-// FIXME: can these constraints be relaxed?
-unsafe impl<T: Send> Send for Deque<T> {}
-unsafe impl<T: Send> Sync for Deque<T> {}
-
-/// Worker half of the work-stealing deque. This worker has exclusive access to
-/// one side of the deque, and uses `push` and `try_pop` method to manipulate it.
-///
-/// There may only be one worker per deque, and operations on the worker
-/// require mutable access to the worker itself.
-#[derive(Debug)]
-pub struct Worker<T> {
```

Comment:
I don't think moving this into a separate crate is a good idea. Either we restrict crossbeam to be purely a collection of primitives and move _all_ structures (queues, stack) out or we keep all structures. I'm leaning towards the latter, because the end result would be a batteries-included low-level concurrency crate containing both primitives and data structures rather than just something people have to build upon to end up with anything useful.